### PR TITLE
Java async client for minio #397

### PR DESCRIPTION
**Pull request by mmadoo**
A first version of async upload (#397) 
If the direction is good, I can do some improvement to use CompletableFuture instead of OnCloseCallback for instance and add documentation.

### CODE REVIEW

File: api/src/main/java/io/minio/MinioClient.java [hunk 1]
```diff
@@ -119,6 +123,9 @@
  * }</pre>
  */
 public class MinioClient extends S3Base {
+	
+  private ExecutorService executor;
```
Comment (balamurugana) [inline api/src/main/java/io/minio/MinioClient.java:127]: My recommendation is to keep async as separate client i.e. `AsyncMinioClient`. I am not sure this is the one we wanted in async support. I did something similar long back https://github.com/balamurugana/minio-java/commit/d6f5878c7d4100eafb0d7ae59506e1f0e7b6e0b1 but we would need to have full support from bottom to top i.e. from `OkHttpClient` to `MinioClient`.

**Comment by mmadoo** [comment]
Hello,
Is it possible to have a feedback on this pull request ?
It have handle all TODO.

**Comment by harshavardhana** [comment]
@mmadoo you can run `./gradlew check` to fix all the things in one shot before you push to this PR.  I see that you are fixing things after you submit the PR and see failure.

**Comment by mmadoo** [comment]
> @mmadoo you can run `./gradlew check` to fix all the things in one shot before you push to this PR. I see that you are fixing things after you submit the PR and see failure.

Sorry,
It should be OK.
```
./gradlew check
./gradlew build
./gradlew runFunctionalTest
```
are succesfull on my computer.

**Comment by mmadoo** [comment]
I have just squashed and rebased.
Test and build are localy successful.

**Comment by balamurugana** [comment]
It is better to have async support like AWS SDK Java. I am looking into that and will assess how much changes in the current code.

**Comment by minio-trusted** [comment]
merged in #1308

**Pull request closed without merging**